config/7.3.0/fapi1part2adv/ig/scripts/groovy/ErrorResponseFilter.groovy

```diff
@@ -33,9 +33,9 @@ return next.handle(context, request).thenAsync {
                 } catch (e){
                     logger.debug("{} error casting json body to be ErrorResponse. Exception: {}" SCRIPT_NAME, e)
                 }
-                return response
+                return newResultPromise(response)
```

Comment:
this return is a from a `Promse.then` , the Function passed should return the plain result and not wrap it in a Promise. You only need to return a `Promise.thenAsync`

`Promise.then` takes a `org.forgerock.util.Function` as an argument, interface docs (note return type is VOUT):
```
public interface Function<VIN, VOUT, E extends Exception> {
    /**
     * Applies this function to the input parameter {@code value} and returns
     * the result.
     *
     * @param value
     *            The input parameter.
     * @return The result of applying this function to {@code value}.
     * @throws E
     *             If this function cannot be applied to {@code value}.
     */
    VOUT apply(VIN value) throws E;
```

thenAsync takes an AsyncFunction, note return type Promise<VOUT>:
```
@FunctionalInterface
public interface AsyncFunction<VIN, VOUT, E extends Exception> {

    /**
     * Asynchronously applies this function to the input parameter {@code value}
     * and returns a {@link Promise} for the result.
     *
     * @param value The input parameter.
     * @return The {@link Promise} representing the result of applying this
     *         function to {@code value}.
     * @throws E If this function cannot be applied to {@code value}, an exception can be thrown instead of returned as
     *           as failed promise.
     */
    Promise<VOUT, E> apply(VIN value) throws E;
```

The second change in this file is correct because that Response is being returned from within thenAsync, therefore it needs to be wrapped in a Promise.

Reply:
Thanks Dan. Of course, that makes complete sense.